src/nccl_ofi_rdma.c

```diff
@@ -5698,11 +6125,33 @@ static inline int get_ep(nccl_net_ofi_device_t *base_dev, nccl_net_ofi_ep_t **ba
  unlock:
 	nccl_net_ofi_mutex_unlock(&device->ep_lock);
 
+	/* During plugin initialization, an endpoint is
+	 * created. Consequently, consider maximum write inline size
+	 * retrieval code as being executed on initialization path the
+	 * first time, avoiding data race on
+	 * `max_write_inline_size`. */
```

Comment:
1. Having this code in `get_ep` seems wrong to me. Why not in `rdma_init`, since we create an endpoint during init time anyway?
2. Technically, we should hold a lock when checking `max_write_inline_size`, since `get_ep` can be called by multiple threads. If you do this during init time instead, that problem goes away.

Reply:
3. The control flow isn't quite right here. Previous steps have `goto unlock` in error cases, at which point you will proceed to execute the `if (max_write_inline_size == 0) ...` code here and overwrite the return code. That would also be resolved by moving this code to init time.

Reply:
4. Sorry, OK I see the comment now about the errors with opening and closing an endpoint in `get_properties`, which is why you want to do this here instead.

I think the right place to do this is in [`nccl_ofi_ofiutils_init_connection`](https://github.com/aws/aws-ofi-nccl/blob/7a0fa83daad9d01c14b2895439e1dd5615cb6ebd/src/nccl_ofi_ofiutils.c#L250-L251). That is where we do similar properties checking/config for OFI endpoints.

Reply:
Where do we retrieve properties in `nccl_ofi_ofiutils_init_connection`? I don't know, `nccl_ofi_ofiutils_init_connection` doesn't seem to be the right place--we don't want to retrieve that value for all rails I think, it should not be part of the SENDRECV path.

> Technically, we should hold a lock when checking max_write_inline_size, since get_ep can be called by multiple threads. If you do this during init time instead, that problem goes away.
I implemented it such that line `max_write_inline_size = size;` is only executed during protocol initialization. 
* Case `get_inject_rma_size_opt` result is `0`: line is never executed, default value of `0` is used.
* Case `get_inject_rma_size_opt` result is not `0`: Whole block of code is not executed due to check `if (max_write_inline_size == 0) {`
As a result, `max_write_inline_size` is only read outside of the initialization path, so there will never be a race on it.

If you want, I can pull in into the lock though, and use lock when `get_properties()` is invoked though for the same of clarity!!

> The control flow isn't quite right here. Previous steps have goto unlock in error cases, at which point you will proceed to execute the if (max_write_inline_size == 0) ... 
That is indeed a problem. Let me know what you think about the points above and I then fix it in the next revision.

Reply:
> Where do we retrieve properties in nccl_ofi_ofiutils_init_connection? I don't know, nccl_ofi_ofiutils_init_connection doesn't seem to be the right place--we don't want to retrieve that value for all rails I think, it should not be part of the SENDRECV path.

OK, I see, you only want to execute this code in RDMA path only. I would have preferred having it somewhere like `ep_rail_init` instead of `get_ep`, but that's splitting hairs.

This is also being fussy, but I dislike that, if `max_write_inline_size == 0`, then we'll keep calling `get_inject_rma_size_opt` every time we create an endpoint. Would prefer giving an initial value `max_write_inline_size == -1`, so we distinguish between `get_inject_rma_size_opt` having never been called, and `get_inject_rma_size_opt` returning not supported or returning 0.

I see your point about locking now. Yeah, I don't think an extra lock is needed.

Reply:
You're right! Using 0 as indicator is rather implicit. Let me change that.